File: src/services/docker/registryClient.ts
import { parseImageReference } from "./imageReference";

export type RegistryCheckStatus = "exists" | "missing" | "unknown";

export interface RegistryCheckResult {
  image: string;
  status: RegistryCheckStatus;
  registry: string;
  repository: string;
  reference: string;
  error?: string;
  suggestion?: string;
}

export interface RegistryClient {
  checkImageExists(image: string, opts?: { signal?: AbortSignal }): Promise<RegistryCheckResult>;
}

export interface RegistryClientOptions {
  fetch?: typeof globalThis.fetch;
  timeoutMs?: number;
}

const MANIFEST_ACCEPT = [
  "application/vnd.oci.image.index.v1+json",
  "application/vnd.oci.image.manifest.v1+json",
  "application/vnd.docker.distribution.manifest.list.v2+json",
  "application/vnd.docker.distribution.manifest.v2+json",
  "application/vnd.docker.distribution.manifest.v1+json",
].join(", ");

interface BearerChallenge {
  realm: string;
  service?: string;
  scope?: string;
}

function resultBase(image: string) {
  const ref = parseImageReference(image);
  return {
    image: ref.original,
    registry: ref.registry,
    repository: ref.repository,
    reference: ref.reference,
  };
}

function registryBaseUrl(registry: string): string {
  return `https://${registry}`;
}

function parseBearerChallenge(value: string | null): BearerChallenge | null {
  if (!value?.startsWith("Bearer ")) return null;
  const params = value.slice("Bearer ".length);
  const parsed: Record<string, string> = {};
  for (const match of params.matchAll(/([a-zA-Z]+)="([^"]*)"/g)) {
    parsed[match[1] ?? ""] = match[2] ?? "";
  }
  if (!parsed.realm) return null;
  return {
    realm: parsed.realm,
    ...(parsed.service ? { service: parsed.service } : {}),
    ...(parsed.scope ? { scope: parsed.scope } : {}),
  };
}

async function requestBearerToken(
  challenge: BearerChallenge,
  fetchImpl: typeof globalThis.fetch,
  signal?: AbortSignal,
): Promise<string | null> {
  const url = new URL(challenge.realm);
  if (challenge.service) url.searchParams.set("service", challenge.service);
  if (challenge.scope) url.searchParams.set("scope", challenge.scope);

  const response = await fetchImpl(url, signal ? { signal } : {});
  if (!response.ok) return null;
  const body = (await response.json()) as { token?: string; access_token?: string };
  return body.token ?? body.access_token ?? null;
}

function createAbortSignal(
  timeoutMs: number,
  signal?: AbortSignal,
): {
  signal?: AbortSignal;
  cleanup: () => void;
} {
  if (timeoutMs <= 0) return signal ? { signal, cleanup: () => {} } : { cleanup: () => {} };
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  const abort = () => controller.abort();
  signal?.addEventListener("abort", abort, { once: true });
  return {
    signal: controller.signal,
    cleanup: () => {
      clearTimeout(timer);
      signal?.removeEventListener("abort", abort);
    },
  };
}

function manifestUrl(image: string): string {
  const ref = parseImageReference(image);
  return `${registryBaseUrl(ref.registry)}/v2/${ref.repository}/manifests/${encodeURIComponent(
    ref.reference,
  )}`;
}

function tagsUrl(image: string): string {
  const ref = parseImageReference(image);
  return `${registryBaseUrl(ref.registry)}/v2/${ref.repository}/tags/list?n=100`;
}

function formatStatus(response: Response): string {
  return `${response.status}${response.statusText ? ` ${response.statusText}` : ""}`;
}

function suggestTag(image: string, tags: string[]): string | undefined {
  const ref = parseImageReference(image);
  if (ref.referenceType !== "tag") return undefined;
  const match = /^(\d+)(.*)$/.exec(ref.reference);
  if (!match)
    return tags.includes("latest") ? ref.original.replace(/:[^/:]+$/, ":latest") : undefined;
  const requestedMajor = Number(match[1]);
  const suffix = match[2] ?? "";
  const candidates = tags
    .map((tag) => {
      const candidate = /^(\d+)(.*)$/.exec(tag);
      if (!candidate || candidate[2] !== suffix) return null;
      return { major: Number(candidate[1]), tag };
    })
    .filter((candidate): candidate is { major: number; tag: string } => candidate !== null)
    .filter((candidate) => candidate.major < requestedMajor)
    .sort((a, b) => b.major - a.major);
  const best = candidates[0]?.tag;
  if (!best) return undefined;
  return `${ref.original.replace(/(:[^/:@]+)?(@.+)?$/, "")}:${best}`;
}

async function fetchTags(
  image: string,
  fetchImpl: typeof globalThis.fetch,
  token: string | null,
  signal?: AbortSignal,
): Promise<string[]> {
  const headers: Record<string, string> = {};
  if (token) headers.Authorization = `Bearer ${token}`;
  const response = await fetchImpl(tagsUrl(image), {
    headers,
    ...(signal ? { signal } : {}),
  });
  if (!response.ok) return [];
  const body = (await response.json()) as { tags?: unknown };
  return Array.isArray(body.tags)
    ? body.tags.filter((tag): tag is string => typeof tag === "string")
    : [];
}

export function createRegistryClient(options: RegistryClientOptions = {}): RegistryClient {
  const fetchImpl = options.fetch ?? globalThis.fetch;
  const timeoutMs = options.timeoutMs ?? 10_000;

  return {
    async checkImageExists(image, opts = {}) {
      const base = resultBase(image);
      const timeout = createAbortSignal(timeoutMs, opts.signal);
      try {
        let token: string | null = null;
        let response = await fetchImpl(manifestUrl(image), {
          method: "HEAD",
          headers: { Accept: MANIFEST_ACCEPT },
          ...(timeout.signal ? { signal: timeout.signal } : {}),
        });

        if (response.status === 401) {
          const challenge = parseBearerChallenge(response.headers.get("www-authenticate"));
          if (!challenge) {
            return { ...base, status: "unknown", error: "registry requires unsupported auth" };
          }
          token = await requestBearerToken(challenge, fetchImpl, timeout.signal);
          if (!token) return { ...base, status: "unknown", error: "registry token request failed" };
          response = await fetchImpl(manifestUrl(image), {
            method: "HEAD",
            headers: { Accept: MANIFEST_ACCEPT, Authorization: `Bearer ${token}` },
            ...(timeout.signal ? { signal: timeout.signal } : {}),
          });
        }

        if (response.status === 405) {
          response = await fetchImpl(manifestUrl(image), {
            method: "GET",
            headers: {
              Accept: MANIFEST_ACCEPT,
              ...(token ? { Authorization: `Bearer ${token}` } : {}),
            },
            ...(timeout.signal ? { signal: timeout.signal } : {}),
          });
        }

        if (response.ok) return { ...base, status: "exists" };
        if (response.status === 404) {
          const tags = await fetchTags(image, fetchImpl, token, timeout.signal);
          const suggestion = suggestTag(image, tags);
          return {
            ...base,
            status: "missing",
            error: "manifest not found",
            ...(suggestion ? { suggestion } : {}),
          };
        }
        if (response.status === 401 || response.status === 403) {
          return { ...base, status: "unknown", error: "registry requires authentication" };
        }
        return { ...base, status: "unknown", error: `registry returned ${formatStatus(response)}` };
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        return { ...base, status: "unknown", error: message };
      } finally {
        timeout.cleanup();
      }
    },
  };
}

File: src/services/docker/registry_client.py
"""Async Docker registry client for image manifest checks.

Parity: ``src/services/docker/registryClient.ts``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, replace
from typing import Any, Literal
from urllib.parse import urlencode

import httpx

from src.services.docker.image_reference import ImageReference, parse_image_reference

RegistryCheckStatus = Literal["exists", "missing", "unknown"]


class RegistryCheckStatusValues:
    """String constants for registry check status (test / validator convenience)."""

    EXISTS: RegistryCheckStatus = "exists"
    MISSING: RegistryCheckStatus = "missing"
    UNKNOWN: RegistryCheckStatus = "unknown"


@dataclass(frozen=True)
class RegistryCheckResult:
    image: str
    status: RegistryCheckStatus
    registry: str
    repository: str
    reference: str
    error: str | None = None
    suggestion: str | None = None


MANIFEST_ACCEPT = (
    "application/vnd.docker.distribution.manifest.v2+json,"
    "application/vnd.docker.distribution.manifest.list.v2+json,"
    "application/vnd.oci.image.index.v1+json,"
    "application/vnd.oci.image.manifest.v1+json"
)


@dataclass
class _BearerChallenge:
    realm: str
    service: str | None = None
    scope: str | None = None


def _parse_bearer_challenge(header: str) -> _BearerChallenge | None:
    if not header.startswith("Bearer "):
        return None
    rest = header[len("Bearer ") :]
    parts: dict[str, str] = {}
    for match in re.finditer(r'(\w+)="([^"]+)"', rest):
        parts[match.group(1)] = match.group(2)
    realm = parts.get("realm")
    if not realm:
        return None
    return _BearerChallenge(
        realm=realm, service=parts.get("service"), scope=parts.get("scope")
    )


async def _request_bearer_token(
    challenge: _BearerChallenge, client: httpx.AsyncClient
) -> str | None:
    params: dict[str, str] = {}
    if challenge.service:
        params["service"] = challenge.service
    if challenge.scope:
        params["scope"] = challenge.scope
    url = challenge.realm
    if params:
        sep = "&" if "?" in url else "?"
        url += sep + urlencode(params)
    response = await client.get(url)
    if response.status_code != 200:
        return None
    body = response.json()
    token = body.get("token") or body.get("access_token")
    return str(token) if token is not None else None


def _manifest_url(registry: str, repository: str, reference: str) -> str:
    return f"https://{registry}/v2/{repository}/manifests/{reference}"


def _tags_url(registry: str, repository: str) -> str:
    return f"https://{registry}/v2/{repository}/tags/list?n=100"


def _suggest_tag(image: str, tags: list[str]) -> str | None:
    ref = parse_image_reference(image)
    if ref.reference_type != "tag":
        return None
    if "latest" in tags:
        base = image.rsplit(":", 1)[0] if ":" in image else image
        return f"{base}:latest"
    return None


class RegistryClient:
    def __init__(self, client: httpx.AsyncClient | None = None, timeout_ms: int = 10_000) -> None:
        self._client = client or httpx.AsyncClient(timeout=timeout_ms / 1000.0)

    async def check_image_exists(
        self, image: str, *, signal: Any | None = None
    ) -> RegistryCheckResult:
        ref = parse_image_reference(image)
        base = RegistryCheckResult(
            image=ref.original,
            status="unknown",
            registry=ref.registry,
            repository=ref.repository,
            reference=ref.reference,
        )
        headers = {"Accept": MANIFEST_ACCEPT}
        token: str | None = None

        try:
            url = _manifest_url(ref.registry, ref.repository, ref.reference)
            response = await self._client.head(url, headers=headers)

            if response.status_code == 401:
                challenge = _parse_bearer_challenge(
                    response.headers.get("WWW-Authenticate", "")
                )
                if challenge is None:
                    return replace(base, error="registry requires unsupported auth")
                token = await _request_bearer_token(challenge, self._client)
                if token is None:
                    return replace(base, error="registry token request failed")
                auth_headers = {**headers, "Authorization": f"Bearer {token}"}
                response = await self._client.head(url, headers=auth_headers)

            if response.status_code == 405:
                auth_headers = (
                    {**headers, "Authorization": f"Bearer {token}"} if token else headers
                )
                response = await self._client.get(url, headers=auth_headers)

            if response.is_success:
                return replace(base, status="exists")
            if response.status_code == 404:
                suggestion = await self._suggest_tag(ref)
                return replace(
                    base,
                    status="missing",
                    error="manifest not found",
                    suggestion=suggestion,
                )
            if response.status_code in (401, 403):
                return replace(base, error="registry requires authentication")
            return replace(
                base,
                error=f"registry returned {response.status_code} {response.reason_phrase}",
            )
        except Exception as err:  # noqa: BLE001
            return replace(base, error=str(err))

    async def _suggest_tag(self, ref: ImageReference) -> str | None:
        url = _tags_url(ref.registry, ref.repository)
        try:
            response = await self._client.get(url)
            if not response.is_success:
                return None
            tags = response.json().get("tags", [])
            if not isinstance(tags, list):
                return None
            str_tags = [t for t in tags if isinstance(t, str)]
            return _suggest_tag(f"{ref.original}", str_tags)
        except Exception:  # noqa: BLE001
            return None


def create_registry_client(
    client: httpx.AsyncClient | None = None, timeout_ms: int = 10_000
) -> RegistryClient:
    return RegistryClient(client=client, timeout_ms=timeout_ms)


__all__ = [
    "RegistryCheckResult",
    "RegistryCheckStatus",
    "RegistryCheckStatusValues",
    "RegistryClient",
    "create_registry_client",
]